infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvironmentVariablesSorter.java

```diff
@@ -0,0 +1,202 @@
+   *     expansion is impossible
+   */
+  static List<EnvVar> sort(Map<String, EnvVar> envVars) throws InfrastructureException {
+    try {
+      EnvironmentVariablesSorter sorter = new EnvironmentVariablesSorter(envVars);
+      List<EnvVar> ret = new ArrayList<>(envVars.values());
+      ret.sort(sorter);
+      return ret;
+    } catch (CycleException e) {
+      throw new InfrastructureException(e.getMessage());
+    }
+  }
+
+  @Override
+  public int compare(EnvVar a, EnvVar b) throws CycleException {
+    if (a == b) {
+      return 0;
+    }
+
+    SortedSet<String> aRefs = referenceMap.get(a.getName());
+    SortedSet<String> bRefs = referenceMap.get(b.getName());
+
+    // quickly deal with the common case of no references in the value
+    if (aRefs.isEmpty()) {
+      // we need to have a total ordering, so if there is no difference in references, we need
+      // to pick a different sorting criteria
+      return bRefs.isEmpty() ? a.getName().compareTo(b.getName()) : -1;
+    } else if (bRefs.isEmpty()) {
+      // b has no references, a has, therefore a > b.
+      return 1;
+    }
+
+    // ok, we have variable references in both a and b
+    // we need to establish the transitive closure over the references to figure out whether there
+    // is a cycle in the dependencies, or not
+    SortedSet<String> aTransRefs = transitiveClosure(a.getName());
+    SortedSet<String> bTransRefs = transitiveClosure(b.getName());
+
+    // we don't need to check for cycle here, because that has been detected when establishing
+    // the closure.
+    if (aTransRefs.contains(b.getName())) {
+      // a depends on b, hence a > b.
+      return 1;
+    } else if (bTransRefs.contains(a.getName())) {
+      // b depends on a -> a < b.
+      return -1;
+    }
+
+    // ok, the two env vars don't depend on each other, so we actually don't care too much about
+    // the order. We need to be consistent and total, though...
+
+    int sizeDiff = aTransRefs.size() - bTransRefs.size();
+    if (sizeDiff != 0) {
+      return sizeDiff;
+    }
+
+    // ah, the two env vars have the same number of references, so we need to take a "slow path"
+    // in determining the total order.
+    Iterator<String> aIt = aRefs.iterator();
+    Iterator<String> bIt = bRefs.iterator();
+    while (aIt.hasNext()) {
+      if (!bIt.hasNext()) {
+        // a has more references than b
+        return 1;
+      }
+
+      String aRef = aIt.next();
+      String bRef = bIt.next();
+
+      int res = aRef.compareTo(bRef);
+      if (res != 0) {
+        return res;
+      }
+    }
+
+    return 0;
+  }
+
+  private SortedSet<String> transitiveClosure(String variableName) {
+    SortedSet<String> recRefs = recursiveReferenceMap.get(variableName);
+    if (recRefs == null) {
+      recRefs = new TreeSet<>(naturalOrder());
+      fillRecursiveReferences(variableName, referenceMap, recRefs);
+      recursiveReferenceMap.put(variableName, recRefs);
```

Comment:
Would `recursiveReferenceMap` -> `recursiveReferencesMap` or maybe `recursiveReferenceMap` -> `transitiveClosures` help? I don't see how this could be simplified apart from renaming.